lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/IO/Resource.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# HLINT ignore "Use void" #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+-- |
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- Implementation of a 'Resource' (think REST) which can be initialized.
+module Cardano.Wallet.Deposit.IO.Resource
+    ( Resource
+    , withResource
+    , ErrResourceMissing (..)
+    , onResource
+    , ErrResourceExists (..)
+    , putResource
+    , ResourceStatus (..)
+    , readStatus
+    ) where
+
+import Prelude
+
+import Control.Concurrent
+    ( forkFinally
+    )
+import Control.Concurrent.Class.MonadMVar
+    ( MonadMVar (..)
+    , putMVar
+    , takeMVar
+    )
+import Control.Concurrent.Class.MonadSTM
+    ( MonadSTM (..)
+    , TVar
+    , atomically
+    , readTVar
+    , writeTVar
+    )
+import Control.Monad
+    ( void
+    )
+import Control.Monad.Class.MonadThrow
+    ( MonadThrow (..)
+    , SomeException
+    )
+
+{-----------------------------------------------------------------------------
+    Resource
+------------------------------------------------------------------------------}
+
+-- | Mutable resource (think REST) that holds a reference of type @a@
+-- that has to be initialized with a 'with…' function.
+data Resource a = Resource
+    { content :: TVar IO (ResourceStatus a)
+    , waitForEndOfLife :: IO ()
+    -- ^ Wait until the 'Resource' is out of scope.
+    }
+
+-- | Possible status of the content of a 'Resource'.
+data ResourceStatus a
+    = NotInitialized
+    | Initializing
+    | Initialized a
+    | Vanished SomeException
+    deriving (Show)
+
+instance Functor ResourceStatus where
+    fmap _ NotInitialized = NotInitialized
+    fmap _ Initializing = Initializing
+    fmap f (Initialized a) = Initialized (f a)
+    fmap _ (Vanished e) = Vanished e
+
+-- | Read the status of a 'Resource'.
+readStatus :: Resource a -> IO (ResourceStatus ())
+readStatus resource = void <$> readTVarIO (content resource)
+
+-- | Make a 'Resource' that can be initialized later.
+--
+-- Once the 'Resource' has been initialized,
+-- it will also be cleaned up once the 'withResource' function has finished.
+--
+-- If the 'Resource' vanishes because of an exception,
+-- the 'withResource' will /not/ be interrupted.
+-- You can use 'getStatus' to poll the current status.
+withResource
```

Comment:
I wonder if we should not put `withResource` in the separate module and package - seems like very reusable thing. Does not `local cluster` use similar approach?